### PR TITLE
CI pipeline that builds and tests the base images

### PR DESCRIPTION
**Pull request by rjanalik**

### CODE REVIEW

File: ci/ci-generator.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#!/usr/local/bin/python
```
Comment (finkandreas) [inline ci/ci-generator.py]: Could we change this to `#!/usr/bin/env python`, then it runs also on systems that do not have python in `/usr/local/bin`?

File: ci/ci-generator.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+         curr = jinja_env.from_string(prev).render(**values)
+         if curr != prev:
+             prev = curr
+         else:
+             return curr
+
+def generate_job(template, os, osver, spackver, target, cudaver=None, rocmver=None):
+    if target == "cpu":
+        spack_target = "alps-zen2"
+    elif target == "cuda":
```
Comment (finkandreas) [inline ci/ci-generator.py]: how does this map to `alps-gh200`?

File: ci/ci-generator.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        spack_target = "alps-a100"
+    elif target == "rocm":
+        spack_target = "alps-mi200"
```
Comment (finkandreas) [inline ci/ci-generator.py:30]: How will this map to `alps-mi300` in the future?

File: ci/cscs.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  - remote: 'https://gitlab.com/cscs-ci/recipes/-/raw/master/templates/v2/.ci-ext.yml'
+
+stages:
+  - configure
+  - run
+
+pipeline-configure:
+  stage: configure
+  extends: .container-runner-lightweight-zen2
+  image: docker.io/rjanalik/ci-generator:python-3.12.3
```
Comment (finkandreas) [inline ci/cscs.yml]: It's fine to use this image, but could you add the Dockerfile to this repository, so it could be understood what is inside?

**Review by finkandreas — CHANGES_REQUESTED** [review]

File: ci/ci-generator.py (hunk 2 above)
Comment (rjanalik) [reply ci/ci-generator.py]: I split `cuda` to `cuda-x86_64` and `cuda-aarch64`.

File: ci/ci-generator.py (hunk 3 above)
Comment (rjanalik) [reply ci/ci-generator.py:30]: The `spack_target` is needed only for tests. I think we can think about this once we have some `mi300`s. Then I could deal with `rocm` the same way as I do now with `cuda`, split it in more targets.

File: ci/cscs.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
+include:
+  - remote: 'https://gitlab.com/cscs-ci/recipes/-/raw/master/templates/v2/.ci-ext.yml'
+
+stages:
+  - prepare
+  - configure
+  - run
+
+variables:
+  CI_GENERATOR_IMAGE_NAME: $CSCS_REGISTRY_PATH/public/ci-generator:$CI_GENERATOR_TAG
```
Comment (finkandreas) [inline ci/cscs.yml]: I would suggest to have this `CI_GENERATOR_IMAGE_NAME: $CSCS_REGISTRY_PATH/public/ci-generator`

File: ci/cscs.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+build-ci-generator-image:
+  stage: prepare
+  extends: .container-builder-cscs-zen2
+  before_script:
+    - echo $CI_GENERATOR_IMAGE_NAME
+    - export CI_GENERATOR_TAG=`cat ci/docker/* | sha256sum - | head -c 16`
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/public/ci-generator:$CI_GENERATOR_TAG
```
Comment (finkandreas) [inline ci/cscs.yml]: with the above change this would be `PERSIST_IMAGE_NAME=$CI_GENERATOR_IMAGE_NAME:$CI_GENERATOR_TAG`

File: ci/cscs.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    - echo "CI_GENERATOR_TAG=$CI_GENERATOR_TAG" > build.env
+  artifacts:
+    reports:
+      dotenv: build.env
+  variables:
+    DOCKERFILE: ci/docker/Dockerfile
+    PERSIST_IMAGE_NAME: $CI_GENERATOR_IMAGE_NAME
```
Comment (finkandreas) [inline ci/cscs.yml]: `PERSIST_IMAGE_NAME: $CI_GENERATOR_IMAGE_NAME` is useless, it's overwritten in the `before_script` anyway.

File: ci/cscs.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+pipeline-configure:
+  stage: configure
+  extends: .container-runner-lightweight-zen2
+  image: $CI_GENERATOR_IMAGE_NAME
```
Comment (finkandreas) [inline ci/cscs.yml]: with above suggestion this would be `$CI_GENERATOR_IMAGE_NAME:$CI_GENERATOR_TAG`

File: ci/cscs.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  before_script:
+    - echo $CI_GENERATOR_IMAGE_NAME
+    - export CI_GENERATOR_TAG=`cat ci/docker/* | sha256sum - | head -c 16`
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/public/ci-generator:$CI_GENERATOR_TAG
+    - echo "CI_GENERATOR_TAG=$CI_GENERATOR_TAG" > build.env
+  artifacts:
+    reports:
+      dotenv: build.env
+  variables:
+    DOCKERFILE: ci/docker/Dockerfile
```
Comment (finkandreas) [inline ci/cscs.yml]: maybe `ci/docker/Dockerfile.ci-generator`?

**Review by finkandreas — COMMENTED** [review]
Some suggestion for better consistency for `CI_GENERATOR_IMAGE_NAME`

**Review by finkandreas — APPROVED** [review]
LGTM

**Pull request merged**